### PR TITLE
Enabling passwordless setup (certificates) and reverse RQ setup (tedp auto connects to te)

### PR DESCRIPTION
**Pull request by aravindhank11**
## To address: https://github.com/vmware/te-ns/issues/9

Issue Detail:
-------------
* TEDP deployed by TE Controller and would need password based authentication to work
* In Apple's testbed -- the SSH is certificate based and so we were unable to deploy Datapath containers.
* Similarly connect is a separate api that was available to connect TE and TEDP, and since this was over SSH and also touches the datastructs internal to TE Controller -- even if we bring up Controller and DP manually the connect won't happen in a passwordless ssh deployments

Logic for TEDP auto connecting to TE Controller -- there by ssh dependency in connect() api disappears
---------------------------------------------------------------------------------------------------------
* docker run command of datapath to take TE Controller IP and flask_port as inputs
* datapath container once comes up, runs a daemon(RQ_CONNECTOR.py) , will make a flask api call get_rq_details() to TE Controller, which would give the RQ details if the datapath is authenticated (refer next pt for how to authenticate). The RQ_CONNETOR.py keeps retrying infinitely to the Controller to authenticate itself.
* A manually deployed datapath container can be authenticated to TE Controller by means of connect() api
    - Internally adds the datapath container IP to setup_completed_tedp set and there by get_rq_details will go through post this.
* Post getting the RQ details, datapath would bring up the stats collector process, establish RQ and monitor them and would re-init them upon failure.
* Once RQ processes are up TEDP makes establish_rq_api() to controller there by controller adds this host to connect_completed_tedps set

Passwordless authentication
----------------------------
* Though the PR doesn't solve the passwordless authentication problem completely, still partially attempts to solve it.
* As a part of docker run command, $HOME/.ssh can be mounted within the container there by system default certs are shared
* If there is no password provided in apis like setup_tedp(), get_cpu_count() and tech_support() which relies on SSH, the API falls back and tries to SSH using system default certificate available in ~/.ssh/
* A separate efforts needs to be done in order to take a PEM file for cert based ssh

### CODE REVIEW

File: te/GET_AND_RUN_DOCKER_IMAGE.py [hunk 1]
```diff
@@ -92,7 +92,7 @@ def getLogger(self):
 class DOCKER_INITIALIZER:
 
     def __init__(self, pathToDocker, ip, nginxPort, stat_collect_interval, stat_dump_interval, \
-        basePath, typeOfDocker, te_controller_ip, te_logpath, te_loglevel):
+        basePath, typeOfDocker, te_controller_ip, te_logpath, te_loglevel, my_ip, flask_port):
```
Comment (aravindhank11) [inline te/GET_AND_RUN_DOCKER_IMAGE.py:95]: GET_AND_RUN_DOCKER_IMAGE.py takes the flask ip and its mgmt_ip as arg if it is called for setup_tedp() -- there by can start the RQ_CONNECTOR.py process

File: te/GET_AND_RUN_DOCKER_IMAGE.py [hunk 2]
```diff
@@ -118,9 +120,11 @@ def __init__(self, pathToDocker, ip, nginxPort, stat_collect_interval, stat_dump
                 'repo_name'           : 'tedp',
                 'run_cmd'             : "docker run --privileged --cap-add=SYS_PTRACE --security-opt " \
                                         "seccomp=unconfined -v /tmp/:/te_host/ -v $HOME:/te_root/ " \
-                                        "-v /var/run/netns:/var/run/netns " \
+                                        "-v $HOME/.ssh/:/root/.ssh/ -v /var/run/netns:/var/run/netns " \
```
Comment (aravindhank11) [inline te/GET_AND_RUN_DOCKER_IMAGE.py:123]: Refer to the PR title where in we try to fallback to password less system default cert based auth if no password is given. To facilitate this we need to mount the ~/.ssh/ folder within the container

File: te/RQ_CONNECTOR.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+# OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+# OF SUCH DAMAGE
+#**********************************************************************************************
+
```
Comment (aravindhank11) [inline te/RQ_CONNECTOR.py:31]: Majority of the code was taken from the removed CONNECTOR.py which was earlier brining up redis queue by talking over SSH.

Logic of RQ_CONNECTOR is described in the PR title

File: te/TE_DP_CONFIG.py [hunk 4]
```diff
@@ -55,13 +55,12 @@ def __init__(self, host_ip, uname, passwd, cpu_count, TEbrokerHandle, lgr, te_me
...
                                             "RESET_DNS"        : self.__result_evaluator_default,
+                                            "CLEAN_TEDP"       : self.__result_evaluator_default,
                                             "UPDATE_DNS"       : self.__result_evaluator_default,
                                             "EXECUTE_CMD"      : self.__result_evaluator_default,
                                             "TECH_SUPPORT"     : self.__result_evaluator_tech_support}
 
-            self.__credentials          = { 'uname':uname,
-                                            'passwd':passwd,
-                                            'host_ip':host_ip }
+            self.__host_ip              = host_ip
```
Comment (aravindhank11) [inline te/TE_DP_CONFIG.py:63]: uname and passwd are only set but never used -- removing the unused variables

File: te/TE_WRAP.py [hunk 5]
```diff
@@ -394,7 +394,10 @@ def __buildDocker(type, image_id=None):
         logging.getLogger("paramiko").setLevel(logging.ERROR)
         ssh = paramiko.SSHClient()
         ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-        ssh.connect(host, username=uname, password=pwd)
+        if pwd:
```
Comment (aravindhank11) [inline te/TE_WRAP.py:397]: where ever password is missing we try to authenticate using system default certs. Auto taken care by paramiko / parallelssh libraries.

In future when we add cert based auth, we can change the priority

File: te/setup_te_swagger.json [hunk 6]
```diff
@@ -30,7 +30,6 @@
                         "name" : "passwd",
                         "description" : "TE controller password",
                         "in" : "query",
-			"required" : true,
```
Comment (aravindhank11) [inline te/setup_te_swagger.json:33]: password field is no longer necessary

File: te/tedp_docker/Dockerfile [hunk 7]
```diff
@@ -109,11 +114,19 @@ RUN tar -xvf ${usr_lib_path}/usr_lib_deps.tar.gz -C / && rm ${usr_lib_path}/usr_
     tar -xvf ${lib64_path}/lib64_deps.tar.gz -C / && rm ${lib64_path}/lib64_deps.tar.gz && \
     ldconfig -v
 
+# add rq.service
```
Comment (aravindhank11) [inline te/tedp_docker/Dockerfile:117]: Adding RQ_CONNECTOR.py as a daemon

File: te/tedp_docker/clean_tedp.sh [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+# OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+# OF SUCH DAMAGE
+#**********************************************************************************************
+
```
Comment (aravindhank11) [inline te/tedp_docker/clean_tedp.sh:33]: clear_config() was happening over ssh as well.
* If remove_container = True --> happens over ssh
* else the rq cleanup happens over rq itself -- where in a job is enqueued to run the script in background (nohup /opt/te/clean_tedp.sh &). This script sleeps for 3 secs there by the RQ job is returned and the cleanup can happen post that.

File: te/TE_WRAP.py [hunk 9]
```diff
@@ -394,7 +397,10 @@ def __buildDocker(type, image_id=None):
...
```
Comment (sudks) [inline te/TE_WRAP.py:400]: Does this logic take care of no-access ( when no ssh related stuff is provided )

File: te/TE_WORK.py [hunk 10]
```diff
@@ -123,36 +127,57 @@ def tech_support(my_ip, remote_ip, remote_user, remote_pwd, remote_path, type_of
...
+        lgr.info("Executing cmd=%s, out=%s, err=%s" %(cmd, out, err))
+
+        try:
+            ssh = paramiko.SSHClient()
+            ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+            if remote_pwd:
+                ssh.connect(remote_ip, username=remote_user, password=remote_pwd)
+            else:
+                ssh.connect(remote_ip, username=remote_user)
+            scp = SCPClient(ssh.get_transport())
```
Comment (sudks) [inline te/TE_WORK.py:174]: Should we have a check if its not ssh enabled systems ( no password, no certs) then we don't even try the SCP logic

File: te/TE_SWAGGER.py [hunk 11]
```diff
@@ -65,7 +65,10 @@ def setup_te():
     #ssh connection to the controller machine
     ssh = paramiko.SSHClient()
     ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-    ssh.connect(controller_ip, username=user, password=password)
+    if password:
+        ssh.connect(controller_ip, username=user, password=password)
+    else:
```
Comment (sudks) [inline te/TE_SWAGGER.py:70]: lets have elif with cert_flag ; else option should just return back

File: te/TE_WRAP.py (hunk 9 above)
Comment (aravindhank11) [reply te/TE_WRAP.py:400]: The function is to deploy TE Controller in write access. One would not use TE_WRAP to deploy controller, but rather go via docker run command

File: te/TE_WORK.py (hunk 10 above)
Comment (aravindhank11) [reply te/TE_WORK.py:174]: If the system is not enabled for ssh, ssh.connect() would raise an exception which is caught in https://github.com/vmware/te-ns/pull/11/files/dd3075eadee0b381e4191045e88146e5446bca2e#diff-3b20c9cd5201723be91202ad245dbb93820190debf2ed38839a7a636c2d7b38dR176-R180

**Review by sudks — APPROVED** [review]

**Pull request merged**